### PR TITLE
Clean up dependencies

### PR DESCRIPTION
**Pull request by munrojm**
- `custodian`, `mpcontribs-client`, `boto3`, and `emmet-core[all]` made optional dependencies
- `AlloyRester` import altered with warning to handle when `pymatgen-analysis-alloys` is not installed via `emmet-core`

### CODE REVIEW

**Comment by munrojm** [comment]
@mkhorton, I have added these warnings on instantiation should `boto3` and `pymatgen-analysis-alloys` not be installed:
```
 UserWarning: Alloy addon package not installed. To query alloy data install the pymatgen-analysis-alloys package.
 UserWarning: boto3 not installed. To query charge density data install the boto3 package.
```
 I also raise an `MPRestError` with the `get_charge_density_from_material_id` method if it is called without having `boto3`.

Furthermore, I have added a custom `__getattr__` method which is called when python cannot resolve an attribute (e.g. `MPRester.alloys` and `MPRester.charge_density`), and have included the same warning messages above as `MPRestError` exceptions.

Want to make sure we are on the same page before I merge anything. Let me know if you have any thoughts.

File: mp_api/client/client.py
Comment (tschaume) [inline mp_api/client/client.py:137]: Separate `ImportError` exception from other exceptions caused by calling `Client()` (e.g. bad api key)

File: mp_api/client/client.py
Comment (tschaume) [inline mp_api/client/client.py:191]: Include URLs to PyPI for optional dependencies?

File: mp_api/client/routes/__init__.py
Comment (tschaume) [inline mp_api/client/routes/__init__.py:39]: Looks like this warning will appear every time the client is initialized without `pymatgen-analysis-alloys` being installed even if the user isn't interested in alloy data. Maybe this could be moved to the part of the code that tries to retrieve alloy data?

File: mp_api/client/routes/__init__.py
Comment (tschaume) [inline mp_api/client/routes/__init__.py:46]: See above re: `AlloysRester`

**Review by tschaume — CHANGES_REQUESTED** [review]
@munrojm looks good overall. Left a few minor comments for your consideration.

File: mp_api/client/client.py
Comment (munrojm) [reply mp_api/client/client.py:137]: Done

File: mp_api/client/client.py
Comment (munrojm) [reply mp_api/client/client.py:191]: Instead of the URL, I have just added the pip command directly.

File: mp_api/client/routes/__init__.py
Comment (munrojm) [reply mp_api/client/routes/__init__.py:39]: Yeah, I debated adding a warning on instantiation. I may just remove this...

**Review by tschaume — APPROVED** [review]

**Comment by munrojm** [comment]
@mkhorton, I am going to merge this unless you have an issue with anything.

**Pull request merged**
